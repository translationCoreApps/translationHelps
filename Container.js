const api = window.ModuleApi;
const React = api.React;
const ReactBootstrap = api.ReactBootstrap;
const View = require('./View');
const fs = require(window.__base + 'node_modules/fs-extra');

const NAMESPACE = 'TranslationHelps';

class Container extends React.Component {
  constructor(){
    super();
    this.state = {
      modalVisibility: false,
    }
    this.wordList = api.getDataFromCheckStore(NAMESPACE, 'wordList');
    this.sectionList = api.getDataFromCheckStore(NAMESPACE, 'sectionList');
    this.followLink = this.followLink.bind(this);
  }
  convertToMarkdown(src) {
    if (!src) return src;
    let replaced = src.replace(/(=+)([^=]+)\1/g, function(match, equals, header) {
        switch(equals.length) {
            case 6:
                return "##" + header;
            case 5:
                return "####" + header;
            default:
                return "#####" + header;
        }
    });
    let followLink = this.followLink;
    window.followLink = followLink;
    return replaced.replace(/\[([^\]]+)\]\(([^)]+)\)/g, '<a href="$2" onclick="return followLink(\'$2\')">$1</a>');

  }

  followLink(word) {
    var currentWord = word.split('/');
    let currentWordFile = currentWord[currentWord.length-1].split('.')[0]+ '.txt';
    var found = null;
    for (var i in this.wordList) {
      if (this.wordList[i].name.trim() == currentWordFile.trim()) {
        found = this.wordList[i].file;
      }
    }
    if (!found) {
        currentWordFile = this.sectionList[currentWord[currentWord.length-1]];
        if (currentWordFile) {
          found = currentWordFile.file;
          let title = found.match(/title: .*/)[0].replace('title: ', '');
          found = found.replace(/---[\s\S]+---/g, '');
          found = '## ' + title + '\n' + found;
        } else {
          try {
            found = fs.readFileSync(__dirname + '/static/words/' + currentWord[currentWord.length-1]).toString();
          } catch(err) {
            console.log(this.props.online ? "You are online" : "You are offline");
            console.log("Found = ", found);
            if(!this.props.online && !found){
              this.setState({modalVisibility: true, modalView: "You are attempting to load an external resource in offline mode, please enable online mode to view this resource"});
              return false;
            }
            console.log(err);
            return true;
          }
        }
    }
    this.setState({
      modalVisibility: true,
      modalView: found
    });
    return false;
  }

  render() {
    let { currentFile} = this.props;
    currentFile = this.convertToMarkdown(currentFile)
    let modalView = this.convertToMarkdown(this.state.modalView)
    return (
        <View
            modalVisibility={this.state.modalVisibility}
            currentFile={currentFile}
            modalFile={modalView|| currentFile}
            showModal={() => this.setState({ modalVisibility: true })}
            hideModal={() => this.setState({ modalVisibility: false, modalView: null })}
        />
    );
  }
}

module.exports = {
    name: NAMESPACE,
    container: Container
}
