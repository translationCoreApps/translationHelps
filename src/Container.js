import React from 'react';
import PropTypes from 'prop-types';
import {ipcRenderer} from 'electron';
import View from './components/View';
import { convertMarkdownLinks } from './helpers/tHelpsHelpers';

class Container extends React.Component {
  constructor(){
    super();
    this.state = {
      modalVisibility: false
    };
    this.followLink = this.followLink.bind(this);
  }

  followLink(link) {
    let linkParts = link.split('/'); // Could be a relative article link, e.g. en/tw/kt/inchrist, or a full URL, e.g. http://unfoldingword.org

    // If the link doesn't have 4 parts, like a relative article link should, or has http:, https: or ftp: treat it as an external link
    if (linkParts.length != 4 || ['http:', 'https:', 'ftp:'].includes(linkParts[0]) ) {
      if (!this.props.online) {
        let message = 'You are attempting to load an external resource in offline mode, please enable online mode to view this resource';
        this.setState({modalVisibility: true, modalView: message});
      } else ipcRenderer.send('open-helper', link);
    } else {
      const [languageId, type, articleCat, articleId] = linkParts;
      let resourceDir = type;
      switch (type) {
        case 'ta':
          resourceDir = 'translationAcademy';
          break;
        case 'tw':
          resourceDir = 'translationWords';
          break;
        default:
          break;
      }

      this.props.actions.loadResourceArticle(resourceDir, articleId, languageId, articleCat);
      let articleData = this.props.resourcesReducer.translationHelps[resourceDir][articleId];

      if (articleData) {
        this.setState({
          modalVisibility: true,
          modalView: articleData,
          articleCat: articleCat
        });
      } else {
        this.setState({
          modalVisibility: true,
          modalView: 'Cannot find specified file.'
        });
      }
    }
  }

  render() {
    const languageId = this.props.projectDetailsReducer.currentProjectToolsSelectedGL[this.props.toolsReducer.currentToolName];
    const followLink = this.followLink;
    const currentFile = convertMarkdownLinks(this.props.currentFile, languageId);
    const modalView = convertMarkdownLinks(this.state.modalView, languageId, this.state.articleCat);
    window.followLink = followLink;
    return (
        <View
            modalVisibility={this.state.modalVisibility}
            currentFile={currentFile}
            modalFile={modalView || currentFile}
            showModal={() => this.setState({ modalVisibility: true })}
            hideModal={() => this.setState({ modalVisibility: false, modalView: null })}
        />
    );
  }
}

Container.propTypes = {
  translate: PropTypes.func,
  actions: PropTypes.shape({
    loadResourceArticle: PropTypes.func.isRequired
  }),
  currentFile: PropTypes.string,
  online: PropTypes.bool,
  projectDetailsReducer: PropTypes.shape({
    currentProjectToolsSelectedGL: PropTypes.object.isRequired
  }),
  resourcesReducer: PropTypes.shape({
    translationHelps: PropTypes.object.isRequired
  }),
  toolsReducer: PropTypes.shape({
    currentToolName: PropTypes.string.isRequired
  }),
};

export default Container;
