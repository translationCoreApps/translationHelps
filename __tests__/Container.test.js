/* eslint-env jest */
import React from 'react';
import Container from '../src/Container';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const props = {
  translate: k=>k,
  actions: {
    loadResourceArticle: jest.fn((lang, type, cat, article)=>{
      return lang+', '+type+', '+cat+', '+article;
    })
  },
  currentFile: '# test #\n\nThis is a [test](rc://en/ta/man/translate/test)',
  online: true,
  projectDetailsReducer: {
    currentProjectToolsSelectedGL: {'testTool': 'en'}
  },
  resourcesReducer: {
    translationHelps: {'type': {
      'validArticle': 'Valid Article Content'
    }}
  },
  toolsReducer: {
    currentToolName: 'testTool'
  },
};

describe('Container Tests', () => {
  it('Test Container', () => {
    const wrapper = shallow(<Container {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Test Container.followLink() with valid article', () => {
    const wrapper = shallow(<Container {...props} />);
    const container = wrapper.instance();
    const link = 'lang/type/cat/validArticle';
    const [lang, type, category, article] = link.split('/');
    const state = container.followLink(link);
    expect(props.actions.loadResourceArticle).toHaveBeenCalled();
    expect(props.actions.loadResourceArticle).toBeCalledWith(type, article, lang, category);
    expect(state).toEqual({
      articleCategory: 'cat',
      modalView: 'Valid Article Content',
      modalVisibility: true
    });
  });

  it('Test Container.followLink() with invalid article', () => {
    const wrapper = shallow(<Container {...props} />);
    const container = wrapper.instance();
    const link = 'lang/type/cat/invalidArticle';
    const [lang, type, category, article] = link.split('/');
    const state = container.followLink(link);
    expect(props.actions.loadResourceArticle).toHaveBeenCalled();
    expect(props.actions.loadResourceArticle).toBeCalledWith(type, article, lang, category);
    expect(state).toEqual({
      articleCategory: 'cat',
      modalView: 'Cannot find an article for lang/type/cat/invalidArticle',
      modalVisibility: true
    });
  });
});
