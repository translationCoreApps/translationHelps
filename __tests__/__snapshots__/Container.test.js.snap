// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Tests Test Container 1`] = `
<View
  currentFile="# test #

This is a <a style=\\"cursor: pointer\\" onclick=\\"return followLink('en/ta/translate/test')\\">test</a>"
  hideModal={[Function]}
  modalFile="# test #

This is a <a style=\\"cursor: pointer\\" onclick=\\"return followLink('en/ta/translate/test')\\">test</a>"
  modalVisibility={false}
  showModal={[Function]}
/>
`;
